svm/src/transaction_processor.rs

```diff
@@ -229,6 +232,32 @@ impl<FG: ForkGraph> TransactionBatchProcessor<FG> {
         );
         load_time.stop();
 
+        let preexecution_account_states: Option<HashMap<Pubkey, AccountSharedData>> =
+            preexecution_account_states_enabled.then(|| {
+                // filter all the writable accounts so that we can notify geyser preexecution account states
+                // this will enable us to track changes in the accounts via geyser
+                // this feature is only enabled if a geyser plugin return true for `enable_pre_trasaction_execution_accounts_data`
+                loaded_transactions
+                    .iter()
+                    .zip(sanitized_txs.iter())
+                    .filter_map(|(transactions, sanitized_transaction)| {
+                        transactions.0.as_ref().ok().map(|transaction| {
+                            transaction.accounts.iter().enumerate().filter_map(
+                                |(account_index, account)| {
+                                    if sanitized_transaction.message().is_writable(account_index) {
+                                        Some(account)
+                                    } else {
+                                        None
+                                    }
+                                },
+                            )
+                        })
+                    })
+                    .flatten()
+                    .cloned()
+                    .collect()
+            });
+
```

Comment:
I think that @buffalojoec had some ideas here about how to get the prior account state without making the SVM transaction processor responsible for doing this work. I think it had something to do with:

1. Getting the `AccountLoader` to accumulate the pristine state of accounts in some cache (it has to load them anyway).
2. Returning that as part of the `ExecutedTransaction`, [just like `programs_modified_by_tx` is.](https://github.com/anza-xyz/agave/blob/e114e3bc4f12d6da687344dc8eb85e36738ef9c5/svm/src/transaction_execution_result.rs#L24)

In this way, you might even be able to _not_ include an account if it's not touched? Could you even omit the properties that weren't dirtied (eg. only return `lamports` if nothing other than `lamports` changed – saving that data from having to be sent to the plugin over the FFI boundary at all)?

Reply:
Yeah, that's pretty much what I had in mind. We just have to make sure we're not doing too much cloning, but my thought is we can maybe consume the `AccountLoader` instance (if it isn't already) when the batch finishes processing, and vend all of the initial account states from that consumed cache.